### PR TITLE
drivers/ninaw10: Connect to WiFi asynchronously.

### PR DESCRIPTION
**Pull request by iabdalkader**
TL;DR: Make WiFi connect non-blocking, to match the behavior of other WiFi modules and drivers.

Before this patch, WiFi connection was blocking, and could raise exceptions if the connection failed for any reason (including timeouts). This doesn't match the behavior of other WiFi modules, which connect asynchronously, and requires handling of exceptions on connect. This change makes `connect()` work asynchronously by scheduling code to poll connection status, and handle reconnects (if needed), and return immediately without blocking.

The following now works without changes on both CYW43 and Nina:

```python
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.connect(WIFI_SSID, WIFI_PASS)

while not wlan.isconnected():
    print("Trying to connect....")
    time.sleep_ms(500)
```

### CODE REVIEW

File: drivers/ninaw10/nina_wifi_drv.h [hunk 1]
```diff
@@ -85,6 +98,8 @@ typedef int (*nina_scan_callback_t)(nina_scan_result_t *, void *);
 
 int nina_init(void);
 int nina_deinit(void);
+int nina_connection_status();
+int nina_connection_reason();
```
Comment (dpgeorge) [inline drivers/ninaw10/nina_wifi_drv.h:102]: please add `void`

File: extmod/network_ninaw10.c [hunk 2]
```diff
@@ -99,6 +103,36 @@ STATIC void network_ninaw10_poll_sockets(mp_sched_node_t *node) {
...
+    if (status == NINA_STATUS_CONNECTED) {
+        // Connected to AP, nothing else to do.
+        return;
+    }
+
+    if (status != NINA_STATUS_NO_SSID_AVAIL) {
+        // If not connected, and no connection in progress, the connection attempt has failed.
+        // Read the ESP failure reason, reconnect and reschedule the connection polling code.
+        int reason = nina_connection_reason();
+        if (reason == 2 || reason == 4 || reason == 6 || reason == 15 || reason >= 200) {
```
Comment (dpgeorge) [inline extmod/network_ninaw10.c:119]: can you use constants for these reasons, instead of these magic numbers?

**Comment by iabdalkader** [comment]
@dpgeorge updated.

File: extmod/network_ninaw10.c [hunk 3]
```diff
@@ -99,6 +103,40 @@ STATIC void network_ninaw10_poll_sockets(mp_sched_node_t *node) {
...
+
+    if (status != NINA_STATUS_NO_SSID_AVAIL) {
+        // If not connected, and no connection in progress, the connection attempt has failed.
+        // Read the ESP failure reason, reconnect and reschedule the connection polling code.
+        int reason = nina_connection_reason();
+        if (reason == NINA_ESP_REASON_AUTH_EXPIRE ||
+            reason == NINA_ESP_REASON_ASSOC_EXPIRE ||
+            reason == NINA_ESP_REASON_NOT_AUTHED ||
+            reason == NINA_ESP_REASON_4WAY_HANDSHAKE_TIMEOUT ||
+            reason >= NINA_ESP_REASON_BEACON_TIMEOUT) {
```
Comment (dpgeorge) [inline extmod/network_ninaw10.c:123]: I can't see these constants defined anywhere, and I don't think this code is built by the CI... might be worth adding `ARDUINO_NANO_RP2040_CONNECT` to the rp2 CI to build this nanw10 driver.

Comment (iabdalkader) [reply extmod/network_ninaw10.c:123]: Yes you're right, they were in `nina_wifi_drv.h` forgot to add it to the commit. 

> might be worth adding ARDUINO_NANO_RP2040_CONNECT to the rp2 CI to build this nanw10 driver.

Would be nice, and it uses other features not used by other boards so would be a good test.

Comment (iabdalkader) [reply extmod/network_ninaw10.c:123]: Done, I think.

**Comment by dpgeorge** [comment]
Thanks for updating.  I also added a line to `tools/ci.sh` to build the Arduino Nano (the line you added just pulled in the submodules for it).
